File: src/my_py_pkg/my_py_pkg/add_two_ints_client.py
#!/usr/bin/env python3
import rclpy
from rclpy.node import Node

from example_interfaces.srv import AddTwoInts
from functools import partial # usado para passar mais paramentros para future.add_done_callback

class AddTwoInstClientNode(Node):
    def __init__(self):
        super().__init__("add_two_ints_client")
        self.call_add_two_ints_server(6, 7)

    # funcao q criara o cliente 
    def call_add_two_ints_server(self, a, b):
        # criando o cliente
        client =  self.create_client(AddTwoInts, "add_two_ints")

        # enquanto o servidor nao responde
        while not client.wait_for_service(1.0):
            self.get_logger().warn("Waiting for Server Add Two Ints")
        
        # criando um objeto de request
        request =  AddTwoInts.Request()
        request.a = a
        request.b = b

        # manda o request de forma assicrona para a variavel future q recebera a resposta
        future = client.call_async(request)
        # calllback para o future
        future.add_done_callback(partial(self.callback_call_add_two_ints, a=a,b=b))
    
    # função que processará a resposta
    def callback_call_add_two_ints(self,future, a, b):
        try:  # recebe a resposta do service e print a resposta
            response = future.result()
            self.get_logger().info(str(a) + " + " + str(b) + " = " + str(response.sum))
            
        except Exception as e:
            self.get_logger().error("Service call failed %r" % (e,))


def main(args=None):
    rclpy.init(args=args)
    node = AddTwoInstClientNode() 
    rclpy.spin(node)
    rclpy.shutdown()


if __name__ == "__main__":
    main()

File: src/my_cpp_pkg/src/add_two_ints_client.cpp
#include "rclcpp/rclcpp.hpp"
#include "example_interfaces/srv/add_two_ints.hpp"
    
class AddTwoIntsClientNode : public rclcpp::Node 
{
public:
    AddTwoIntsClientNode() : Node("add_two_ints_client") 
    {   
        // passando para thread a função
        thread1 = std::thread(std::bind(&AddTwoIntsClientNode::callAddTwoIntsService, this, 1, 4));
        thread2 = std::thread(std::bind(&AddTwoIntsClientNode::callAddTwoIntsService, this, 5, 10));

        // ou
        threads.push_back(std::thread(std::bind(&AddTwoIntsClientNode::callAddTwoIntsService, this, 1, 4)));
        threads.push_back(std::thread(std::bind(&AddTwoIntsClientNode::callAddTwoIntsService, this, 5, 10)));
    }

    // função para chamar o serviço
    void callAddTwoIntsService(int a, int b)
    {
        // criando um client
        auto client = this->create_client<example_interfaces::srv::AddTwoInts>("add_two_ints");
        // mandando uma mensagem enquanto o server nao responde
        while(!client->wait_for_service(std::chrono::seconds(1)))
        {
            RCLCPP_WARN(this->get_logger(),  "Waiting for the server to be up...");
        }

        // criando request e passando os numeros
        auto request = std::make_shared<example_interfaces::srv::AddTwoInts::Request>();
        request->a = a;
        request->b = b;

        // resposta recebida do server
        auto future = client->async_send_request(request);

        // processando a resposta
        try{
            auto response = future.get();
            RCLCPP_INFO(this->get_logger(), "%ld + %ld = %ld", request->a, request->b, response->sum);
        }
        catch(const std::exception &e){
            RCLCPP_ERROR(this->get_logger(),  "Service call failed");
        }
    }
    
private:
    // criando uma thread para rodar a função pois se ela for chamada direto pelo construtor
    // o nó não executará o spin
    std::thread thread1;
    std::thread thread2;
    std::vector<std::thread> threads;
};
    
int main(int argc, char **argv)
{
    rclcpp::init(argc, argv);
    auto node = std::make_shared<AddTwoIntsClientNode>();
    rclcpp::spin(node);
    rclcpp::shutdown();
    return 0;
}